avocado/core/test.py

```diff
@@ -116,40 +127,37 @@ class Test(unittest.TestCase):
     default_params = {}
 
     def __init__(self, methodName='test', name=None, params=None,
-                 base_logdir=None, tag=None, job=None, runner_queue=None):
+                 base_logdir=None, job=None, runner_queue=None):
         """
         Initializes the test.
 
         :param methodName: Name of the main method to run. For the sake of
                            compatibility with the original unittest class,
                            you should not set this.
-        :param name: Pretty name of the test name. For normal tests, written
-                     with the avocado API, this should not be set, this is
-                     reserved for running random executables as tests.
+        :param name: Pretty name of the test name. For normal tests,
+                     written with the avocado API, this should not be
+                     set.  This is reserved for internal Avocado use,
+                     such as when running random executables as tests.
+        :type name: :class:`avocado.core.test.TestName`
         :param base_logdir: Directory where test logs should go. If None
                             provided, it'll use
                             :func:`avocado.data_dir.create_job_logs_dir`.
-        :param tag: Tag that differentiates 2 executions of the same test name.
-                    Example: 'long', 'short', so we can differentiate
-                    'sleeptest.long' and 'sleeptest.short'.
         :param job: The job that this test is part of.
+        :raises: :class:`avocado.core.test.NameNotTestNameError`
         """
         def record_and_warn(*args, **kwargs):
             """ Record call to this function and log warning """
             if not self.__log_warn_used:
                 self.__log_warn_used = True
             return original_log_warn(*args, **kwargs)
 
-        _incorrect_name = None
-        if isinstance(name, basestring):    # TODO: Remove in release 0.37
-            _incorrect_name = True
-            self.name = TestName(0, name)
-        elif name is not None:
+        if name is not None:
+            if not isinstance(name, TestName):
+                raise NameNotTestNameError(name)
```

Comment:
TBH, I'd prefer to not raise anything.  We're using Python and we should not be in the business of doing type checking.  I only added the exception because it's a deprecation (type change) that may be worth being flagged more explicitly and because your deprecation comment made reference to that behavior.

And since we're trying to be explicit, yes, I think it's worth going all the way and having a custom Exception, because it makes more clear what's going on.